src/apps/cainteoir.man
.TH CAINTEOIR 1 "FEBRUARY 2014" "Cainteoir Text-to-Speech" "Cainteoir Manuals"
.SH NAME
cainteoir \- listen to and record electronic books and documents
.SH SYNOPSIS
.B cainteoir [OPTION..]
.I DOCUMENT
.SH DESCRIPTION
.B cainteoir
is a command\-line interface for Cainteoir Text-to-Speech.

The voice used to speak the document is determined in the following
order: the voice passed on the command-line; the language passed on
the command-line; the language specified by the document. If no
voice is found, the default voice is used.
.SH OPTIONS
.IP "-c, --contents"
List the table of contents for the specified document.
.IP "-D DEVICE, --device=DEVICE"
Use the DEVICE ALSA/pulseaudio device name for audio output, e.g.
when the default device does not work.
.IP "-f FROM, --from=FROM"
Start reading/recording from the specified table of content entry.
If not specified, this defaults to the start of the document. The
value is the number provided from the table of contents output.
.IP "-h, --help"
Show a command-line option usage help message.
.IP "-l LANG, --language=LANG"
Select a text-to-speech voice that can speak in the specified
language.
.IP "-M, --metadata"
Output the RDF metadata for the Cainteoir Text-to-Speech engine
and supported voices. This is output as RDF N-Triple statements.
.IP "-m, --monotone"
Set the voice to monotone. Equivalent to setting the pitch range
to 0.
.IP "--narrator"
Use any embedded audio found in the document.
.IP "-o FILE, --output=FILE"
The file to output to when recording audio. If not specified,
the author and title are used to create a file in the current
directory.
.IP "-p PITCH, --pitch=PITCH"
Set the voice's base pitch.
.IP "-P RANGE, --pitch-range=RANGE"
Set the voice's pitch variance to the specified value (with 0
being monotone).
.IP "-q, --quiet"
Don't display the playing/recording time.
.IP "-r [FORMAT], --record[=FORMAT]"
Record the document to the specified audio format ('wav' for wave
audio, 'ogg' for Ogg/Vorbis audio). If no format is specified,
wave audio is generated.
.IP "-s SPEED, --speed=SPEED"
Set the voice's reading speed to the specified number of words
per minute.
.IP "--stdout"
When recording, write the audio data to standard output instead
of a file.
.IP "-t TO, --to=TO"
Read/record upto the specified table of content entry. If not
specified, this defaults to the end of the document. The value
is the number provided from the table of contents output.
.IP "--tts-fallback"
Use the currently selected TTS voice when there is no embedded
audio for the current text.
any embedded audio found in the document.
.IP "-v VOICE, --voice=VOICE"
Select the text-to-speech voice with the specified name.
.IP "-V VOLUME, --volume=VOLUME"
Set the voice's volume to the specified value.
.SH AUDIO BOOKS
With supported audio books -- such as ePub 3.0 documents with Media
Overlays -- Cainteoir Text-to-Speech allows you to use the audio within
that document instead of the Text-to-Speech voice. There are 3 possible
modes of playback:
.IP "text-to-speech only"
In this mode, any embedded audio is ignored and the selected TTS voice
is used to speak the text.
.IP "narration only"
In this mode (using the
.B --narrator
option), only the embedded audio of the document is used.
.IP "narration with text-to-speech fallback"
In this mode (using the
.B --narrator --tts-fallback
options), the embedded audio of the document is used if available,
otherwise the currently selected TTS voice is used.
.SH PLAYER COMMANDS
These commands are available when listening to or recording a document:
.IP "q"
Stops the reading or recording and exits the application.
.SH AUTHOR
Reece H. Dunn <msclrhd@gmail.com>
.SH "SEE ALSO"
.BR espeak (1)


NEWS

-------------------------------------------------------------------------------
VERSION 0.8 - Friday June 1st 2012 [Jun 2012]

  *  define an archive interface with a zip file implementation
  *  report the filename as a toc entry if no toc is present
  *  move the UTF-8 helpers to a common location
  *  optimize decoding of utf-8 and us-ascii encodings
  *  do not load a new decoder when switching to the same encoding
  *  switch document parsing from an event model to a reader model

applications:
  *  doc2doc: a simple command-line tool for converting documents from
     one format to another

document formats:
  *  pdf:     support PDF documents using the Poppler library
  *  (x)html: support extracting metadata from several meta tag names
  *  zip:     support reading generic zipped document collections
  *  ePub:    make the ePub parser use the archive interface

-------------------------------------------------------------------------------
VERSION 0.7 - Friday March 30th 2012 [Mar 2012]

  *  use BCP 47 for interpreting and comparing language tags
  *  translate the language and region names using the iso-codes package
  *  support for espeak using installed mbrola voices

document formats:
  *  (x)html: report h1..h6 sections as table of content items
  *  email:   improve email content detection and support newsgroup information
  *  rdf/xml: support property attributes on an empty property element
  *  xml:     support entities in DTD blocks

bug fixes:
  *  fix detection of html that is valid xml, but is not xhtml
  *  improve detection of control characters when identifying binary data

cleanup/refactoring:
  *  simplified and cleaned up the RDF model and query APIs
  *  use std::shared_ptr consistently over std::tr1::shared_ptr
  *  use std::make_shared instead of allocating the object explicitly
  *  remove the libxml2 dependency and make all xml-based parsers use xmlreader

-------------------------------------------------------------------------------
VERSION 0.6 -- Sunday January 29th 2012 [Jan 2012]

  *  languages.rdf -- represent ISO-639-2/B, ISO-639-2/T, ISO-639-3, ISO-639-5,
     espeak and wikimedia language codes;
  *  add document format implementation status documentation.

xmlreader xml parsing support:
  *  parse entities in attribues;
  *  support single-quote and html-style attributes;
  *  parse xml namespaces;
  *  unify xml and rdf namespace/uri support;
  *  support element/attribute mapping to context entry objects.

document parsers:
  *  epub  -- support epub 3 @refines and @datatype metadata;
  *  html  -- support all known implicit close tags and recognise more elements;
  *  ocf   -- use the xmlreader api and expose a reader api;
  *  opf   -- use the xmlreader api.

bug fixes:
  *  release the file handle when finished recording ogg files.

-------------------------------------------------------------------------------
VERSION 0.5 -- Monday November 21st 2011 [Nov 2011]

  *  use the shared-mime-info database for MIME type detection.
  *  use iconv for character encoding conversions.
  *  support all Content-Transfer-Encoding types for MIME headers, including base64.
  *  improve audio error handling during reading/recording.
  *  support more language code mappings for new espeak voices.
  *  support UND as a language identifier (for Calibri eBooks).
  *  support XML encoded HTML without an associated xmlns.

-------------------------------------------------------------------------------
VERSION 0.4 -- Tuesday July 26th 2011 [Jul 2011]

file formats:
  *  support ncx-based table of content in epub documents;
  *  support epub 3.0 metadata;
  *  basic support for SSML.

cainteoir application and engine:
  *  support listing a document's table of contents and read/record a
     specified content range;
  *  support setting the voice reading speed, pitch and volume.

-------------------------------------------------------------------------------
VERSION 0.3 -- Friday July 1st 2011 [Jun 2011]

file formats:
  *  support for single-file HTML pages (MHTML, MHT);
  *  support for RTF documents;
  *  enhanced HTML support.

bug fixes:
  *  don't segfault if the OPF document is missing from an epub file.

cainteoir:
  *  improved total reading time estimating heuristics.

-------------------------------------------------------------------------------
VERSION 0.2 -- Sunday May 29th 2011 [May 2011]

file formats:
  *  support email (RFC822) mbox format;
  *  basic support for RTF documents.

metadata:
  *  add rdfs:domain, rdfs:range and OWL markup to the text-to-speech schema;
  *  install the text-to-speech RDF schema used by the Cainteoir Engine;
  *  extract language and title metadata from HTML documents;
  *  extract metadata from email From and Subject headers.

cainteoir:
  *  list supported document and audio formats in the --help output;
  *  don't output progress when writing audio data to stdout;
  *  don't crash when using the default voice (no dc:language metadata).

-------------------------------------------------------------------------------
VERSION 0.1 -- Wednesday April 27th 2011 [Apr 2011]

cainteoir:
  *  command-line text-to-speech program;
  *  support for using espeak text-to-speech voices;
  *  listen to text, html and epub documents on the command line;
  *  record text, html and epub documents to ogg/vorbis and wave audio;
  *  see the reading/recording progress and (estimated) total time;
  *  preserve metadata from document to audio (currently epub to ogg/vorbis).

metadata:
  *  extract metadata from epub, smil, rdf/xml and other documents;
  *  generate RDF N-Triples
  *  generate RDF Turtle statements
  *  generate VorbisComment metadata


src/apps/cainteoir.cpp
/* Cainteoir Command-Line Application.
 *
 * Copyright (C) 2010-2014 Reece H. Dunn
 *
 * This file is part of cainteoir-engine.
 *
 * cainteoir-engine is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * cainteoir-engine is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with cainteoir-engine.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "config.h"
#include "compatibility.hpp"
#include "i18n.h"
#include "options.hpp"

#include <cainteoir/engines.hpp>
#include <cainteoir/document.hpp>
#include <stdexcept>
#include <iostream>
#include <cstdio>
#include <cmath>

#include <termios.h>
#include <unistd.h>
#include <fcntl.h>
#include <stdio.h>
#include <limits.h>

namespace rdf = cainteoir::rdf;
namespace rql = cainteoir::rdf::query;
namespace tts = cainteoir::tts;

enum actions
{
	speak,
	show_metadata,
	show_contents,
};

int termchar()
{
	struct termios oldt, newt;
	int ch;
	int oldf;

	tcgetattr(STDIN_FILENO, &oldt);
	newt = oldt;
	newt.c_lflag &= ~(ICANON | ECHO);
	tcsetattr(STDIN_FILENO, TCSANOW, &newt);

	oldf = fcntl(STDIN_FILENO, F_GETFL, 0);
	fcntl(STDIN_FILENO, F_SETFL, oldf | O_NONBLOCK);

	ch = getchar();

	tcsetattr(STDIN_FILENO, TCSANOW, &oldt);
	fcntl(STDIN_FILENO, F_SETFL, oldf);

	return ch;
}

void format_time(char *s, int n, double seconds)
{
	int ms = int(seconds * 10.0) % 10;

	int minutes = floor(seconds / 60.0);
	seconds = seconds - (minutes * 60.0);

	int hours = floor(minutes / 60.0);
	minutes = minutes - (hours * 60.0);

	snprintf(s, n, "%02d:%02d:%02d.%01d", hours, minutes, (int)floor(seconds), ms);
}

void status_line(double elapsed, double total, double progress, const char *state)
{
	char elapsed_time[80];
	char total_time[80];

	format_time(elapsed_time, 80, elapsed);
	format_time(total_time, 80, total);

#define HIDE_CURSOR "\033[?25l"
#define SHOW_CURSOR "\033[?25h"
	fprintf(stdout, HIDE_CURSOR " : %s of %s [%.2f%%] : %s        \r" SHOW_CURSOR, elapsed_time, total_time, progress, state);
#undef  SHOW_CURSOR
#undef  HIDE_CURSOR

	fflush(stdout);
}

int main(int argc, char ** argv)
{
	setlocale(LC_MESSAGES, "");
	bindtextdomain(PACKAGE, LOCALEDIR);
	textdomain(PACKAGE);

	try
	{
		actions action = speak;
		bool use_narrator = false;
		bool use_tts_fallback = false;
		bool show_progress = true;

		const char *voicename = nullptr;
		const char *language = nullptr;
		const char *outfile = nullptr;
		const char *outformat = nullptr;
		const char *device_name = nullptr;

		int speed = INT_MAX;
		int pitch = INT_MAX;
		int range = INT_MAX;
		int volume = INT_MAX;

		std::pair<size_t, size_t> nav_range = { -1, -1 };

		const option_group general_options = { nullptr, {
			{ 'M', "metadata", bind_value(action, show_metadata),
			  i18n("Show the RDF metadata for the engine and voices") },
			{ 'q', "quiet", bind_value(show_progress, false),
			  i18n("Do not print any output (including current playing/recording time)") },
			{ 'D', "device", device_name, "DEVICE",
			  i18n("Use DEVICE for audio output (ALSA/pulseaudio device name)") },
		}};

		const option_group speech_options = { i18n("Speech:"), {
			{ 'v', "voice", voicename, "VOICE",
			  i18n("Use the voice named VOICE") },
			{ 'l', "language", language, "LANG",
			  i18n("Use a voice that speaks the language LANG") },
			{ 's', "speed", speed, "SPEED",
			  i18n("Set the voice's speed to SPEED words per minute") },
			{ 'p', "pitch", pitch, "PITCH",
			  i18n("Set the voice's base pitch to PITCH") },
			{ 'P', "pitch-range", range, "RANGE",
			  i18n("Set the voice's pitch to vary by RANGE") },
			{ 'V', "volume", volume, "VOLUME",
			  i18n("Set the voice's volume to VOLUME percent") },
			{ 'm', "monotone", bind_value(range, 0),
			  i18n("Set the voice to monotone (pitch varies by 0)") },
		}};

		const option_group narration_options = { i18n("Narration:"), {
			{ 0, "narrator", bind_value(use_narrator, true),
			  i18n("Use any narration/embedded audio in the document") },
			{ 0, "tts-fallback", bind_value(use_tts_fallback, true),
			  i18n("Use the TTS voice if no narration is provided for text") },
		}};

		const option_group toc_options = { i18n("Table of Contents:"), {
			{ 'c', "contents", bind_value(action, show_contents),
			  i18n("List the table of contents for the specified document") },
			{ 'f', "from", nav_range.first, "FROM",
			  i18n("Start reading/recoding from contents marker FROM") },
			{ 't', "to", nav_range.second, "TO",
			  i18n("Finish reading/recording after contents marker TO") },
		}};

		const option_group recording_options = { i18n("Recording:"), {
			{ 'o', "output", outfile, "FILE",
			  i18n("Recorded audio is written to FILE") },
			{ 0, "stdout", bind_value(outfile, "-"),
			  i18n("Recorded audio is written to the standard output") },
			{ 'r', "record", outformat, "FORMAT",
			  i18n("Record the audio as a FORMAT file (default: wav)") },
		}};

		const std::initializer_list<option_group> options = {
			general_options,
			speech_options,
			narration_options,
			toc_options,
			recording_options,
		};

		const std::initializer_list<const char *> usage = {
			i18n("cainteoir [OPTION..] DOCUMENT"),
			i18n("cainteoir [OPTION..]"),
		};

		if (!parse_command_line(options, usage, argc, argv))
			return 0;

		if (nav_range.second != -1) ++nav_range.second;
		if (outformat && !strcmp(outformat, "wave"))
			outformat = "wav";

		auto mode = tts::media_overlays_mode::tts_only;
		if (use_narrator)
			mode = use_tts_fallback
			     ? tts::media_overlays_mode::tts_and_media_overlays
			     : tts::media_overlays_mode::media_overlays_only;

		rdf::graph metadata;
		cainteoir::supportedDocumentFormats(metadata, cainteoir::text_support);
		cainteoir::supportedAudioFormats(metadata);

		cainteoir::tts::engines tts(metadata);

		if (action == show_metadata)
		{
			(*rdf::create_formatter(std::cout, rdf::formatter::turtle))
				<< rdf::rdf
				<< rdf::rdfa
				<< rdf::rdfs
				<< rdf::xsd
				<< rdf::xml
				<< rdf::owl
				<< rdf::dc
				<< rdf::dcterms
				<< rdf::dcam
				<< rdf::epub
				<< rdf::opf
				<< rdf::ocf
				<< rdf::pkg
				<< rdf::media
				<< rdf::ncx
				<< rdf::dtb
				<< rdf::smil
				<< rdf::xhtml
				<< rdf::skos
				<< rdf::foaf
				<< rdf::tts
				<< rdf::iana
				<< rdf::subtag
				<< metadata;
			return 0;
		}

		if (voicename)
		{
			const rdf::uri *ref = tts::get_voice_uri(metadata, rdf::tts("name"), voicename);
			if (ref)
				tts.select_voice(metadata, *ref);
		}
		else if (language)
		{
			const rdf::uri *ref = tts::get_voice_uri(metadata, rdf::dc("language"), language);
			if (ref)
				tts.select_voice(metadata, *ref);
		}

		if (speed  != INT_MAX) tts.parameter(tts::parameter::rate)->set_value(speed);
		if (pitch  != INT_MAX) tts.parameter(tts::parameter::pitch)->set_value(pitch);
		if (range  != INT_MAX) tts.parameter(tts::parameter::pitch_range)->set_value(range);
		if (volume != INT_MAX) tts.parameter(tts::parameter::volume)->set_value(volume);

		const char *filename = (argc == 1) ? argv[0] : nullptr;
		rdf::uri subject(filename ? filename : std::string(), std::string());
		auto reader = cainteoir::createDocumentReader(filename, metadata, std::string());
		if (!reader)
		{
			fprintf(stderr, i18n("unsupported document format for file \"%s\"\n"), filename ? filename : "<stdin>");
			return EXIT_FAILURE;
		}

		cainteoir::document doc(reader, metadata);
		auto listing = cainteoir::navigation(metadata, subject, rdf::epv("toc"));

		if (action == show_contents)
		{
			int toc_number = 1;
			for (auto &entry : listing)
			{
				printf(" %4d ", toc_number);
				for (int i = 0; i < entry.depth; ++i)
					printf("... ");
				printf("%s\n", entry.title.c_str());
				++toc_number;
			}
			return 0;
		}

		std::string author;
		std::string title;
		for (auto &query : rql::select(metadata, rql::subject == subject))
		{
			if (rql::predicate(query).ns == rdf::dc || rql::predicate(query).ns == rdf::dcterms)
			{
				const rdf::uri &object = rql::object(query);
				if (object.empty())
				{
					if (rql::predicate(query).ref == "title")
						title = rql::value(query);
					else if (rql::predicate(query).ref == "creator")
						author = rql::value(query);
				}
				else
				{
					rql::results selection = rql::select(metadata, rql::subject == object);

					if (rql::predicate(query).ref == "creator")
					{
						std::string role;
						std::string value;

						for (auto &data : selection)
						{
							const std::string &object = rql::value(data);
							if (rql::predicate(data) == rdf::rdf("value"))
								value = object;
							else if (rql::predicate(data) == rdf::opf("role") || rql::predicate(data) == rdf::pkg("role"))
								role = object;
						}

						if (!value.empty() && (role == "aut" || role.empty()))
							author = value;
					}
					else if (rql::predicate(query).ref == "title")
					{
						for (auto &data : selection)
						{
							if (rql::predicate(data) == rdf::rdf("value"))
								title = rql::value(data);
						}
					}
				}
			}
		}

		std::shared_ptr<cainteoir::audio> out;
		const char *state;
		if (outformat || outfile)
		{
			state = i18n("recording");
			std::stringstream file;

			if (outfile)
				file << outfile;
			else
				file << author << " - " << title << "." << outformat;

			std::string outfile = file.str();

			if (!outformat)
				outformat = "wav";

			out = cainteoir::create_audio_file(outfile.c_str(), outformat, 0.3, metadata, subject, metadata, tts.voice());
			if (!out.get())
				throw std::runtime_error(i18n("unsupported audio file format"));

			if (outfile != "-")
			{
				fprintf(stdout, i18n("Recording \"%s\"\n"),   filename);
				fprintf(stdout, i18n("       to \"%s\"\n\n"), outfile.c_str());
			}
			else
				show_progress = false;
		}
		else
		{
			state = i18n("reading");
			out = cainteoir::open_audio_device(device_name, metadata, subject, metadata, tts.voice());

			fprintf(stdout, i18n("Reading \"%s\"\n\n"), filename);
		}

		if (show_progress)
		{
			fprintf(stdout, i18n("Author : %s\n"), author.c_str());
			fprintf(stdout, i18n("Title  : %s\n\n"), title.c_str());
		}

		auto speech = tts.speak(out, listing, doc.children(listing, nav_range), mode);
		while (speech->is_speaking())
		{
			if (show_progress)
				status_line(speech->elapsedTime(), speech->totalTime(), speech->completed(), state);

			switch (termchar())
			{
			case 'q':
				speech->stop();
				break;
			default:
				usleep(250);
				break;
			}
		}

		if (show_progress)
		{
			status_line(speech->elapsedTime(), speech->totalTime(), speech->completed(), i18n("stopped"));
			fprintf(stdout, "\n");
		}
	}
	catch (std::runtime_error &e)
	{
		fprintf(stderr, i18n("error: %s\n"), e.what());
		return EXIT_FAILURE;
	}

	return EXIT_SUCCESS;
}
